File: 416 Partition Equal Subset Sum/416 Partition Equal Subset Sum.go
func canPartition(nums []int) bool {
    sum := 0
    for _, n := range nums {
        sum += n
    }
    if sum % 2 == 1 {
        return false
    }

    sort.Slice(nums, func(i, j int) bool { return nums[i] > nums[j] })

    sum /= 2

    dp := make([]int, len(nums))

    for i, n := range nums {
        if n == sum {
            return true
        }
        if n > sum {
            continue
        }
        max := n
        for j := 0; j < i; j++ {
            v := dp[j] + n
            if v == sum {
                return true
            }
            if v < sum && v > max {
                max = v
            }
        }
        dp[i] = max
    }
    return false
}

File: 416 Partition Equal Subset Sum/416 Partition Equal Subset Sum.py
"""
Given a non-empty array nums containing only positive integers, find if the array can be partitioned into two subsets such that the sum of elements in both subsets is equal.

Example 1:

Input: nums = [1,5,11,5]
Output: true
Explanation: The array can be partitioned as [1, 5, 5] and [11].

Example 2:

Input: nums = [1,2,3,5]
Output: false
Explanation: The array cannot be partitioned into equal sum subsets.
"""
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        nums_sum = sum(nums)
        if nums_sum % 2 == 1:
            return False

        target = nums_sum // 2
        dp = 1
        for val in nums:
            dp |= (dp<<val)
        return (dp & (1<<target))
